extensions-contrib/protobuf-extensions/src/main/java/io/druid/data/input/protobuf/ProtoBufInputRowParser.java

```diff
@@ -0,0 +1,138 @@
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.data.input.protobuf;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.annotation.JsonTypeName;
+import com.github.os72.protobuf.dynamic.DynamicSchema;
+import com.google.protobuf.ByteString;
+import com.google.protobuf.Descriptors;
+import com.google.protobuf.Descriptors.Descriptor;
+import com.google.protobuf.DynamicMessage;
+import com.google.protobuf.InvalidProtocolBufferException;
+import com.google.protobuf.util.JsonFormat;
+import io.druid.data.input.ByteBufferInputRowParser;
+import io.druid.data.input.InputRow;
+import io.druid.data.input.MapBasedInputRow;
+import io.druid.data.input.impl.ParseSpec;
+import io.druid.java.util.common.parsers.ParseException;
+import io.druid.java.util.common.parsers.Parser;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.nio.ByteBuffer;
+import java.util.Map;
+
+@JsonTypeName("protobuf")
+public class ProtoBufInputRowParser implements ByteBufferInputRowParser {
+  //private static final Logger log = new Logger(ProtoBufInputRowParser.class);
+
+  private final ParseSpec parseSpec;
+  private Parser<String, Object> parser;
+  private final String descriptorFilePath;
+  private final String protoMessageType;
+  private Descriptor descriptor;
+
+
+  @JsonCreator
+  public ProtoBufInputRowParser(
+      @JsonProperty("parseSpec") ParseSpec parseSpec,
+      @JsonProperty("descriptor") String descriptorFilePath,
+      @JsonProperty("protoMessageType") String protoMessageType
+  ) {
+    this.parseSpec = parseSpec;
+    this.descriptorFilePath = descriptorFilePath;
+    this.protoMessageType = protoMessageType;
+    this.parser = parseSpec.makeParser();
+    this.descriptor = getDescriptor(descriptorFilePath);
+  }
+
+  @Override
+  public ParseSpec getParseSpec() {
+    return parseSpec;
+  }
+
+  @Override
+  public ProtoBufInputRowParser withParseSpec(ParseSpec parseSpec) {
+    return new ProtoBufInputRowParser(parseSpec, descriptorFilePath, protoMessageType);
+  }
+
+  @Override
+  public InputRow parse(ByteBuffer input) {
+    String json;
+    try {
+      DynamicMessage message = DynamicMessage.parseFrom(descriptor, ByteString.copyFrom(input));
+      json = JsonFormat.printer().print(message);
+      // log.trace(json);
+    } catch (InvalidProtocolBufferException e) {
+      throw new ParseException(e, "Protobuf message could not be parsed");
+    }
+
+    Map<String, Object> record = parser.parse(json);
```

Comment:
I guess this means that the parseSpec has to be a "json" parseSpec, since we're parsing protobuf by turning it into json first. This isn't great for performance, but the old code wasn't a paragon of efficiency either, so I'm willing to let it slide.

Although: have you looked into whether it'd be possible to apply the flattenSpec to the DynamicMessage directly? If doable, that should be faster than either this code or the old code.